rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md

```diff
@@ -0,0 +1,555 @@
+        activation='relu')(inputs)
+    outputs = keras.layers.Dense(1, activation='relu')(outputs)
+    model = keras.Model(inputs=inputs, outputs=outputs)
+    model.compile(loss='mse',
+                  metrics=['mae'])
+    # Model training
+    history = model.fit(
+        x,
+        y,
+        epochs=100,
+        validation_data=validation_data,
+        # Tune whether to use shuffle.
+        shuffle=hp.Boolean("shuffle"),
+        # Tune whether to use sample_weights.
+        sample_weight=sample_weight if hp.Boolean("sample_weight") else None,
+        # The provided callbacks list contains checkpointing and tensorboard.
+        callbacks=callbacks)
+    # Save the model to a unique path with `trial_id`.
+    model.save(os.path.join(trial.trial_id, 'model'))
+    # Returning the evaluation results
+    return np.min(history.history["val_mae"])
+
+# When Tuner.run_trial is overridden,
+# `hypermodel` and `objective` are optional.
+tuner = MyTuner(
+    max_trials=3,
+    executions_per_trial=2,
+    overwrite=True,
+    directory="my_dir",
+    project_name="helloworld",
+)
+
+# Anything passed to `search()` will
+# go to `**kwargs` for `Tuner.run_trial()`.
+tuner.search(x, y)
+# Get the best model.
+best_model = tuner.get_best_models()[0]
+```
+
+There are several important features in this workflow:
+
+* Tune the arguments in `HyperModel.fit()`, like `shuffle` and `sample_weight`.
+
+* Share local variables across the workflow. For example, the model building
+  process can access the `num_features`, which is a variable in data
+  preprocessing. It solves the problem of joint tuning.
+
+* Use built-in callbacks for convenience. The callbacks argument contains
+  callback functions for checkpointing and TensorBoard setup.
+
+* The return value is flexible. It can be a single value, or a list of values,
+  or a dictionary of metrics, or even a `History` object returned by
+  `model.fit()`.
+
+* The `hypermodel` and `objective` can be optional. The user doesn't need to
+  define a `HyperModel`. If the return value is a single value, it will be
+  minimized by default. Therefore, objective is also optional.
+
+* The user can build a unique path to save each model with `trial.trial_id`.
+
+For the use case of reusing existing Keras code. The user can use the following
+workflow, which calls a function using all the hyperparameters. The user only
+needs to write a function to call the existing Keras code and return the
+evaluation results.
+
+```py
+class MyTuner(kt.Tuner):
+ def run_trial(self, trial, **kwargs):
+   hp = trial.hyperparameters
+   return build_and_evaluate_model(
+       hp.Int("num_features", 10, 15),
+       hp.Choice('units', [8, 16, 32]),
+       ...
+       trial.trial_id,
+   ))
+   # Save model can be handled by the user.
+   # `trial_id` is unique for each trial.
+
+tuner = MyTuner(...)
+tuner.search()
+# Retraining the model
+build_and_evaluate_model(**tuner.get_best_hyperparameters()[0])
```

Comment:
Might be worth mentioning that `build_and_evaluate_model` is a utility and then appending a brief description about its inputs and what it does.